### ISSUE TITLE
parallelize graph-building

### ISSUE THREAD

**Issue by rkurchin** [aced-differentiate/ChemistryFeaturization.jl#22]
It's a one-time cost for a given dataset, but it's quite slow (~1k/hr) for larger sets, and should be easy to batch out in some reasonably automated way...

**Comment by thazhemadam** [comment]
Do you have something specific in mind, or would something like getting [this loop](https://github.com/aced-differentiate/ChemistryFeaturization.jl/blob/cd905382add821e3e2c4f699107829946e4049a9/src/pmg_graphs.jl#L168-L186) to run on different threads work well enough?

**Comment by rkurchin** [comment]
yeah something like that would be great! I didn't have anything more specific than that in mind, perhaps just adding an argument to specify number of threads (and/or perhaps a flag of whether to parallelize at all and it'll figure out how many there are) to the batch build function...

**Comment by thazhemadam** [comment]
hmm. I'll try and see how we implement at least that much (for now). 
If successful, I'll also try actually benchmarking the "before" and "after" (using something like [BenchmarkTools.jl](https://github.com/JuliaCI/BenchmarkTools.jl)), and have those results also posted too.

**Comment by thazhemadam** [comment]
Few things worth noting wrt implementing this feature by running [this loop](https://github.com/aced-differentiate/ChemistryFeaturization.jl/blob/cd905382add821e3e2c4f699107829946e4049a9/src/pmg_graphs.jl#L168-L186) on different threads (I might be going about these in the wrong way, and if I am, I'd appreciate any and all pointers):
* Running that loop on different threads directly (using `Threads.@threads` or even `Threads.@spawn` ) causes a segmentation fault. This seems to be caused because of [this line](https://github.com/aced-differentiate/ChemistryFeaturization.jl/blob/cd905382add821e3e2c4f699107829946e4049a9/src/pmg_graphs.jl#L35). Might have to try implementing some kind of synchronous locking mechanism that ensures only one thread can call `aseio.read()` at a time, and see how that works out.
*  It might be worth noting that the original `aseio.read()`  (directly from the python package itself) can be run on parallel threads without any issues (I was able to test that out using a Python script).

**Comment by rkurchin** [comment]
Have you tried using Distributed and just a `pmap`? I've had some success with that when parallelizing this stuff "manually" so maybe something about the way that handles stuff internally plays nicer with PyCall? I'm far from an expert on the internals there, but worth a shot.

**Comment by thazhemadam** [comment]
Yes I tried that, and it still results in a segmentation fault. FWIW, I tried two variants of the function called in `pmap` - one called only  `pyimport_conda()`, and the other called only `aseio.read()` (with `aseio` declared as a global variable).
